src/librustc_trans/context.rs

```diff
@@ -843,6 +843,10 @@ impl<'a, 'tcx> LayoutTyper<'tcx> for &'a SharedCrateContext<'a, 'tcx> {
     type TyLayout = TyLayout<'tcx>;
 
     fn layout_of(self, ty: Ty<'tcx>) -> Self::TyLayout {
+        if let Some(&layout) = self.tcx().layout_cache.borrow().get(&ty) {
+            return TyLayout { ty: ty, layout: layout, variant_index: None };
+        }
+
         self.tcx().infer_ctxt((), traits::Reveal::All).enter(|infcx| {
```

Comment:
I'd still like to get rid of the `normalize_associated_type` method - can't `field_of` rely on `layout_of` normalizing before checking the cache? `if !ty.has_projection_types() {` is really fast, right?